proposals/provenance-logging.md

```diff
@@ -0,0 +1,128 @@
+## Motivation
+Reproducible ML is a [big problem](https://petewarden.com/2018/03/19/the-machine-learning-reproducibility-crisis/). 
+Kubeflow could add a lot of values to users by making it easy for users to track how datasets and models were produced.
+A lot of valuable information is available in K8s resource specifications. Resources like K8s Jobs, TFJobs, ConfigMaps, and Argo workflows
+often contain information (e.g. via command line arguments) about inputs and outputs and the code (docker image) used. There are
+two major problems to using this information to track provenance of models and data today:
+
+1. The information is not durable
+
+	* I should not rely on the K8s APIServer as a persistent, historical data store for resources.
+
+1. The schema is not convenient
+
+	* People use a variety of conventions/patterns to pass inputs/outputs to their programs
+	* Lack of uniformity is a barrier to building tools that can surface information in useful ways.
+
+The purpose of this proposal is to solve these problems to facilitate tools to track and surface provenance.
+
+## Goals
+
+1. Build an agent to persist resource specs to durable storage
```

Comment:
Are we building an agent? Or just creating a standard to which all components in the system must act? E.g. spit out any changes to /audit. 

Also, we should enforce that auditing has been written via hashing/etc - e.g. the next step (e.g. training) would not proceed without the previous one (e.g. data transformation) being written and signed.

Reply:
We are building an agent.

There's nothing here that can enforce the next step not starting without the previous one being written and signed. 

That would be the responsibility of an orchestration/pipeline system like Pachyderm. 

/cc @dwhitena